### ISSUE TITLE
Compile / build borgbackup with musl

### ISSUE THREAD

**Issue by noplanman** [borgbackup/borg#5459]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Question

## Describe the question.

Does anybody have experience building borgbackup with musl?

This would be great, to add support for OpenWrt based routers and any other devices that prefer musl instead of glibc.

**Comment by ThomasWaldmann** [comment]
I never tried...

**Comment by noplanman** [comment]
I'm asking because of an issue we got on the ARM binary builder repo:
https://git.feneas.org/borg-binary-builder/borg-binaries/-/issues/6

Would be super cool to have binaries for OpenWrt routers! (not only because I have one too 😉)

As this is completely unknown territory for me, it would be great if somebody with experience would chime in and see if they could get it to work 🙏

**Comment by Gu1nness** [comment]
Seems interesting.
There is a Vagrant VM for OpenWRT for testing, and setting up information
```
[build]
compiler=musl
```
 in [setup.cfg](https://github.com/borgbackup/borg/blob/master/setup.cfg) could do the trick to build with it for this specific VM. What do you think @ThomasWaldmann

**Comment by ThomasWaldmann** [comment]
Don't most OpenWrt based systems have too little RAM for borg anyway?

**Comment by noplanman** [comment]
@ThomasWaldmann How much RAM does Borg need / recommend?

I have a Turris Omnia with 1GB and there are versions with 2GB RAM, which should be more than enough, right?

**Comment by ThomasWaldmann** [comment]
It depends, there are some formulas on the docs to estimate RAM usage.

**Comment by noplanman** [comment]
@Gu1nness I've tried your suggestion, but then end up with this error:
`error: don't know how to compile C/C++ code on platform 'posix' with 'musl' compiler`

I've installed `musl-tools` (which in turn installs `musl` and `musl-dev`).
How can I make python know about the `musl` compiler? Haven't found anything online 😕

**Comment by 2xsaiko** [comment]
I have no problem using the latest commit (fc925878fc6b040d3e5c3997aa46649e0a145c8d) of borg with musl on Gentoo x86-64 so far fwiw. Release 1.1.14 and before crashes with [this error][1] when trying to run it for me, although that might be fixable on Gentoo's side.

[1]: https://paste.2x.ax/~saiko/f131415ad0704d184d53d91b0e8f82b9761a498e

**Comment by ThomasWaldmann** [comment]
@2xsaiko we do not use libgomp, so not sure why there is an issue with that or whether that has anything to do with borg.

**Comment by noplanman** [comment]
@2xsaiko Would you mind sharing what you mean exactly? You're able to build borg from source with `musl`?

If you could share how, I'd hugely appreciate it, thanks 🙏

**Comment by 2xsaiko** [comment]
I didn't do anything special, just installed the app-backup/borgbackup-9999 ebuild (version 9999 ebuilds use the source from the latest git commit by default). I did have to put `export BORG_OPENSSL_PREFIX=/usr` into /etc/portage/env/app-backup/borgbackup-9999 since it needed that variable set when building. This is on a standard gentoo install using the musl profile and with the musl overlay enabled (which patches a bunch of stuff that would otherwise not compile, although I'm not sure it's relevant here).

Or, the non-gentoo specific version (that I tried to test if the error I mentioned above was a bug on gentoo's side or not), I cloned the repo, made a new virtualenv (`python -m venv env`, `. env/bin/activate`) so I didn't have to globally install borg, and then did `BORG_OPENSSL_PREFIX=/usr python setup.py install`. This also works as expected using the latest commit.

I assume you're trying to cross-compile for musl though instead of directly on the target?

**Comment by noplanman** [comment]
> I assume you're trying to cross-compile for musl though instead of directly on the target?

Exactly. But if a virtual target with the required musl setup can be used as well, that would work. I'm just so lost with this that I really don't know what to be looking for 😕 

If there were a docker image I could take that has all musl "stuff" set up, that would be perfect.

**Comment by noplanman** [comment]
I've tried with a bunch of docker images that I found but my know-how is just too limited. Hope somebody with more experience can find a way to build a binary that works with musl, like for OpenWRT-based routers.

**Comment by ThomasWaldmann** [comment]
Looks like there is a borg package for alpine linux, so it seems to work with musl:

https://pkgs.alpinelinux.org/package/edge/community/x86_64/borgbackup
